llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -6871,6 +6974,18 @@ void LoopVectorizationCostModel::collectValuesToIgnore() {
     const SmallVectorImpl<Instruction *> &Casts = IndDes.getCastInsts();
     VecValuesToIgnore.insert(Casts.begin(), Casts.end());
   }
+
+  // Ignore any values that we know will be flattened
+  for (auto It : getPartialReductionChains()) {
+    PartialReductionChain Chain = It.second;
+    SmallVector<Value *> PartialReductionValues{Chain.Reduction, Chain.BinOp,
+                                                Chain.ExtendA, Chain.ExtendB,
+                                                Chain.Accumulator};
+    ValuesToIgnore.insert(PartialReductionValues.begin(),
+                          PartialReductionValues.end());
+    VecValuesToIgnore.insert(PartialReductionValues.begin(),
```

Comment:
Would it be possible to shift the detection of partial reductions to after the initial VPlans have been constructed and work based on VPlan? The pattern matching done in `addPartialReductionIfSupported(` should be possible to do on VPlan and then introducing the partial reductions could be done as VPlan transformation. That would allow the cost to be computed directly on the VPlan (given ::computeCost is implemented for the recipe) and more accurately (VPlan DCE will take care of removing unneeded recipes)

If possible this would be preferable instead of adding additional logic and mappings to the legacy cost model.

Reply:
I'd like that too, but we need to make sure the higher VFs are available to consider. I guess another way to do that would be to make `AArch64TTIImpl::shouldMaximizeVectorBandwidth()` return true for SVE registers as well as NEON registers.

Reply:
I might be missing something, but at least for NEON we seem to already consider VF = 16 for loops that perform dot-products without this change? https://godbolt.org/z/nqqM7Gxvv

It looks like its just that the lowering doesn't make use of dot products for those cases yet

Reply:
NEON uses a VF of 16 because we do return true for `shouldMaximizeVectorBandwidth()` when the register type is fixed length vectors. It doesn't lower to a dot product because we're looking for the partial reduction intrinsic, which isn't being emitted yet.

I wonder if (for now) we can try always maximizing the bandwidth if there's an add reduction in the loop and the target supports partial reductions, then deal with detecting the specifics during vplan construction/transformation.

I didn't see any major swings in spec2K17 performance when enabling max vector bandwidth for SVE, but there might be a build failure in one benchmark since it didn't run with that change in. I'll take a closer look.

Reply:
Just checking back in if this direction proved viable?

Reply:
I believe so, I have https://github.com/llvm/llvm-project/pull/109671 to enable max bandwidth which does allow us to move the code to vplan recipes. There's some concern about whether there will be some performance regressions by doing so, but I wasn't able to find any when running spec2K17, lnt, or a few hpc kernels. So I'm tempted to commit that and fix up any we do find later.

I think the recipe-based transformation for LV can proceed independently of that. It's useful for NEON in the current form, and will be for SVE once we enable max bw there.